### PR TITLE
Add validation-aware type narrowing and taint analysis for `FormRequest`

### LINKED ISSUE #558: taint: add @psalm-taint-source input for Validator::validated() and FormRequest::safe()
## Problem

Validated data is still raw user input. Validation checks format (required, max, email), it does NOT sanitize content. Developers widely assume `$request->validated()` is safe and pipe it directly into SQL, HTML output, or shell commands.

This is the single most dangerous taint gap in the plugin.

## Methods

| Class | Method | Annotation |
|---|---|---|
| `Illuminate\Validation\Validator` | `validated()` | `@psalm-taint-source input` |
| `Illuminate\Validation\Validator` | `validate()` | `@psalm-taint-source input` |
| `Illuminate\Validation\Validator` | `safe()` | `@psalm-taint-source input` |
| `Illuminate\Validation\Validator` | `getData()` | `@psalm-taint-source input` |
| `Illuminate\Validation\Validator` | `getValue($attribute)` | `@psalm-taint-source input` |
| `Illuminate\Validation\Validator` | `attributes()` | `@psalm-taint-source input` |
| `Illuminate\Foundation\Http\FormRequest` | `validated()` | `@psalm-taint-source input` |
| `Illuminate\Foundation\Http\FormRequest` | `safe()` | `@psalm-taint-source input` |

## Stub files to create

- `stubs/common/Validation/Validator.stubphp`
- `stubs/common/Foundation/Http/FormRequest.stubphp`

## ValidatedInput flow annotations

`ValidatedInput` wraps validated data. Flow annotations needed:

- `all()`, `input()`, `toArray()`, `__get()`, `offsetGet()` — `@psalm-flow`
- `only()`, `except()`, `collect()`, `str()`/`string()` — `@psalm-flow`

NOT sources: `integer()`, `float()`, `boolean()` (numeric cast), `enum()` (strict whitelist), `date()` (structured Carbon).

Stub: `stubs/common/Support/ValidatedInput.stubphp`

## Example (undetected today)

```php
public function store(StorePostRequest $request) {
    $data = $request->validated(); // looks safe — it's validated!
    DB::insert('INSERT INTO posts (title) VALUES (' . $data['title'] . ')'); // SQLi
    return response("<h1>{$data['title']}</h1>"); // XSS
}
```

OWASP: A03:2021 Injection

### PR DESCRIPTION
**Pull request by alies-dev**
## What does this PR do?

Adds type narrowing and taint analysis for `FormRequest::validated()` and `FormRequest::safe()` based on declared validation rules.

**Type narrowing:** `$request->validated('age')` with `'age' => 'required|integer'` returns `int|numeric-string` instead of `mixed`. `$request->validated()` returns a typed array shape.

**Taint analysis:** Fields with safe rules (`integer`, `uuid`, `alpha_num`, `date`, `in:...`) have taint removed. Fields with unsafe rules (`string`, `email`) remain tainted. Stubs mark `Validator::validated()`, `FormRequest::safe()`, and `ValidatedInput` methods as `@psalm-taint-source input`.

Closes #558

## How was it tested?

- Type test: `ValidationTypeNarrowingTest.phpt` — single field types, full array shape, nullable, sometimes, wildcards (`items.*`, `items.*.field`)
- Taint tests: `SafeValidatedIntegerNoTaint.phpt` (integer rule removes taint), `TaintedHtmlValidatedString.phpt` (string rule keeps taint)
- Unit test: `ValidationRuleAnalyzerTest.php` — 21 tests covering rule-to-type mapping, taint bitmasks, modifiers, edge cases
- `composer test` all pass, `composer psalm` clean

## Checklist
- [x] Tests cover the change (type test in `tests/Type/` and/or unit test in `tests/Unit/`)
- [ ] Documentation is updated (if needed, otherwise remove this item)

### CODE REVIEW

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+    private static function ruleToType(string $rule, ?string $param): ?Union
+    {
+        return match ($rule) {
+            'string'                           => Type::getString(),
+            'integer'                          => new Union([new TInt(), new TNumericString()]),
+            'numeric'                          => new Union([new TInt(), new TFloat(), new TNumericString()]),
+            // Laravel's boolean rule accepts: true, false, 0, 1, '0', '1'
+            'boolean'                          => Type::combineUnionTypes(
+                Type::getBool(),
+                new Union([new TLiteralInt(0), new TLiteralInt(1)]),
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+     * @psalm-mutation-free
+     */
+    private static function findRulesReturnInClass(Node\Stmt\Class_ $class): ?Node\Expr\Array_
+    {
+        foreach ($class->stmts as $classStmt) {
+            if (!$classStmt instanceof Node\Stmt\ClassMethod) {
+                continue;
+            }
+
+            if ($classStmt->name->toString() !== 'rules') {
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Addressed in previous commit.

File: stubs/common/Support/ValidatedInput.stubphp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * @param  string|null  $tz
+     * @return \Carbon\Carbon|null
+     */
+    public function date($key, $format = null, $tz = null) {}
+
+    /**
+     * @template TEnum of \BackedEnum
+     * @param  string  $key
+     * @param  class-string<TEnum>  $enumClass
+     * @return TEnum|null
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply stubs/common/Support/ValidatedInput.stubphp]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationTaintHandler.php [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return TaintKind::ALL_INPUT;
+    }
+
+    /**
+     * Remove taint from validated('field') calls where the field's validation
+     * rules guarantee safe content (e.g. integer, uuid, alpha_num).
+     */
+    #[\Override]
+    public static function removeTaints(AddRemoveTaintsEvent $event): int
+    {
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationTaintHandler.php:65]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+        return match ($rule) {
+            'integer', 'numeric', 'boolean',
+            'uuid', 'ulid',
+            'alpha', 'alpha_num', 'alpha_dash',
+            'date', 'date_format',
+            'in',
+            'file', 'image'                         => TaintKind::ALL_INPUT,
+            'url', 'active_url',
+            'ip', 'ipv4', 'ipv6'                    => TaintKind::ALL_INPUT & ~TaintKind::INPUT_SSRF,
+            // string, email, json, regex, required, max, min, etc. → keep all taint
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationTaintHandler.php [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            try {
+                // validate() is on Request (parent), validated() is on FormRequest
+                $isRequest = $className === \Illuminate\Http\Request::class
+                    || $className === \Illuminate\Foundation\Http\FormRequest::class
+                    || $codebase->classExtends($className, \Illuminate\Http\Request::class);
+            } catch (\Psalm\Exception\UnpopulatedClasslikeException|\InvalidArgumentException) {
+                continue;
+            }
+
+            if ($isRequest) {
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationTaintHandler.php]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+    {
+        return match ($rule) {
+            'integer', 'numeric', 'boolean',
+            'uuid', 'ulid',
+            'alpha', 'alpha_num', 'alpha_dash',
+            'date', 'date_format',
+            'in',
+            'file', 'image'                         => TaintKind::ALL_INPUT,
+            // string, email, url, ip, json, regex, required, max, min, etc. → keep all taint
+            default                                 => 0,
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationTaintHandler.php [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+        if (!$expr->name instanceof Identifier) {
+            return null;
+        }
+
+        $methodName = \strtolower($expr->name->toString());
+
+        if ($methodName !== 'validated' && $methodName !== 'validate') {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationTaintHandler.php:132]: Addressed in previous commit.

File: stubs/common/Support/ValidatedInput.stubphp [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+     * @param  string|null  $key
+     * @param  bool  $default
+     * @return bool
+     */
+    public function boolean($key = null, $default = false) {}
+
+    /**
+     * @param  string  $key
+     * @param  string|null  $format
+     * @param  \UnitEnum|string|null  $tz
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply stubs/common/Support/ValidatedInput.stubphp:120]: Addressed in previous commit.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        array $callArgs,
+        MethodReturnTypeProviderEvent $event,
+    ): ?Union {
+        $firstArgType = $event->getSource()->getNodeTypeProvider()->getType($callArgs[0]->value);
+
+        if ($firstArgType instanceof Union && $firstArgType->isSingleStringLiteral()) {
+            $key = $firstArgType->getSingleStringLiteral()->value;
+
+            if (isset($rules[$key])) {
+                return $rules[$key]->type;
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php]: Addressed in previous commit.

File: tests/Application/laravel-test.sh [hunk 11]
```diff
@@ -118,6 +118,11 @@ info "Creating a new Laravel project using installer v${LARAVEL_INSTALLER_VERSIO
 composer create-project --quiet --prefer-dist laravel/laravel "$APP_INSTALLATION_PATH" "$LARAVEL_INSTALLER_VERSION"
 cd "$APP_INSTALLATION_PATH"
 
+info "Installing first-party Laravel packages for realistic analysis coverage"
+composer require --quiet laravel/sanctum laravel/scout laravel/pennant laravel/socialite laravel/jetstream laravel/cashier
+./artisan jetstream:install livewire --teams
+./artisan pennant:feature ExampleFeature
+
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply tests/Application/laravel-test.sh]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 12]
```diff
@@ -0,0 +1,663 @@
...
+            if ($parentRule?->nullable) {
+                $listType = Type::combineUnionTypes($listType, Type::getNull());
+            }
+
+            $topLevel[$parent] = new ResolvedRule(
+                $listType,
+                0, // Taint removal applies per-element, not to the list container
+                $parentRule?->nullable ?? false,
+                $parentRule?->sometimes ?? false,
+            );
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php:474]: Addressed in previous commit.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+        foreach ($rules as $field => $resolvedRule) {
+            $fieldType = $resolvedRule->type;
+
+            if ($resolvedRule->sometimes) {
+                $fieldType = $fieldType->setPossiblyUndefined(true);
+            }
+
+            $properties[$field] = $fieldType;
+        }
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:312]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 14]
```diff
@@ -0,0 +1,663 @@
...
+
+        // Default: if no type rule matched, return mixed (don't narrow)
+        $type ??= Type::getMixed();
+
+        // nullable modifier: add null to type union
+        if ($nullable) {
+            $type = Type::combineUnionTypes($type, Type::getNull());
+        }
+
+        return new ResolvedRule($type, $removedTaints, $nullable, $sometimes);
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Addressed in previous commit.

File: tests/Type/tests/ValidationTypeNarrowingTest.phpt [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    // nullable string → null|string
+    $_bio = $request->validated('bio');
+    /** @psalm-check-type-exact $_bio = null|string */
+}
+
+function testValidatedFullShape(StoreUserRequest $request): void
+{
+    $_all = $request->validated();
+    /** @psalm-check-type-exact $_all = array{name: string, age: int|numeric-string, score: float|int|numeric-string, role: 'admin'|'guest'|'user', bio: null|string, nickname?: string} */
+}
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply tests/Type/tests/ValidationTypeNarrowingTest.phpt:60]: Addressed in previous commit.

File: stubs/common/Foundation/Http/FormRequest.stubphp [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * Get the validated data from the request.
+     *
+     * @param  array|int|string|null  $key
+     * @param  mixed  $default
+     * @return mixed
+     *
+     * @psalm-taint-source input
+     */
+    public function validated($key = null, $default = null) {}
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply stubs/common/Foundation/Http/FormRequest.stubphp:16]: Addressed in previous commit.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 17]
```diff
@@ -0,0 +1,683 @@
...
+
+            $parentRule = $topLevel[$parent] ?? null;
+
+            /** @var non-empty-array<string, Union> $properties */
+            $properties = [];
+
+            foreach ($children as $childField => $childRule) {
+                $childType = $childRule->type;
+
+                if ($childRule->sometimes) {
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php]: Fixed — wildcard children now check `!childRule->required` alongside `sometimes`.

File: tests/Type/tests/ValidationTypeNarrowingTest.phpt [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+
+    $_age = $safe->input('age');
+    /** @psalm-check-type-exact $_age = int|numeric-string */
+
+    // str() and string() also narrow via the TRequest template
+    $_nameStr = $safe->str('name');
+    /** @psalm-check-type-exact $_nameStr = string */
+
+    $_nameString = $safe->string('name');
+    /** @psalm-check-type-exact $_nameString = string */
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply tests/Type/tests/ValidationTypeNarrowingTest.phpt]: Fixed — `str()`/`string()` now fall through to stub return type (`Stringable`), only `input()` is narrowed.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 19]
```diff
@@ -0,0 +1,683 @@
...
+            'alpha', 'alpha_num', 'alpha_dash',
+            'hex_color', 'mac_address',
+            'date', 'date_format',
+            'before', 'before_or_equal',
+            'after', 'after_or_equal',
+            'date_equals',
+            'timezone',
+            'in'                                    => TaintKind::ALL_INPUT,
+            // file, image, mimes, mimetypes → keep taint (file names/paths/contents are user-controlled)
+            // string, email, url, ip, json, regex, required, max, min, etc. → keep all taint
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php:223]: Verified safe: `strtotime()` rejects HTML/SQL/shell metacharacters (`<`, `>`, `'`, `;`, `--`). Date strings that pass validation cannot carry injection payloads.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+    {
+        foreach ($union->getAtomicTypes() as $atomic) {
+            if ($atomic instanceof TNamedObject) {
+                /** @psalm-var class-string */
+                return $atomic->value;
+            }
+        }
+
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:230]: Already works — in Psalm 7, `StoreUserRequest&static` is a single `TNamedObject` with `extra_types`. The test at line 85 confirms `ValidatedInput<StoreUserRequest&static>` resolves correctly.

File: src/Handlers/Validation/ValidationRuleAnalyzer.php [hunk 21]
```diff
@@ -0,0 +1,683 @@
...
+            ) {
+                $required = true;
+            }
+
+            // Type-bearing rules (first one wins for type, all accumulate taint)
+            $ruleType = self::ruleToType($ruleName, $ruleParam);
+
+            if ($ruleType instanceof \Psalm\Type\Union && !$type instanceof \Psalm\Type\Union) {
+                $type = $ruleType;
+            }
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationRuleAnalyzer.php:137]: By design — first type-bearing rule wins. Added unit test `first_type_bearing_rule_wins` to lock in behavior.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 22]
```diff
@@ -0,0 +1,315 @@
...
+        $properties = [];
+
+        foreach ($rules as $field => $resolvedRule) {
+            $fieldType = $resolvedRule->type;
+
+            // Fields without a presence rule (required, present, accepted, etc.)
+            // may be absent from validated() output when not provided in the request.
+            if (!$resolvedRule->required) {
+                $fieldType = $fieldType->setPossiblyUndefined(true);
+            }
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:309]: Added: `sometimes_required_field_is_sometimes_and_required`, `rule_order_does_not_affect_taint`, `first_type_bearing_rule_wins` unit tests.

File: src/Handlers/Validation/ValidationTaintHandler.php [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+use Psalm\Type\Union;
+
+/**
+ * Manages taint for validated data from FormRequest and Request.
+ *
+ * When a MethodReturnTypeProvider overrides the return type of validated() or validate(),
+ * Psalm skips the stub's @psalm-taint-source annotation. This handler compensates by
+ * explicitly adding taint (via AddTaintsInterface) and then selectively removing it
+ * for fields with safe rules (via RemoveTaintsInterface).
+ *
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidationTaintHandler.php:24]: Created upstream issue: https://github.com/vimeo/psalm/issues/11765. Added link to ValidationTaintHandler docblock and all known-limitation taint tests.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 24]
```diff
@@ -0,0 +1,315 @@
...
+ *
+ * ValidatedInput is generic: ValidatedInput<TRequest of FormRequest>. When safe() returns
+ * ValidatedInput<static>, the template parameter carries the concrete FormRequest class,
+ * enabling type narrowing on ValidatedInput accessor methods.
+ *
+ * Known limitation: when this handler provides a return type, Psalm skips the stub's
+ * @psalm-taint-source annotation for variable assignments. This means taint is lost
+ * when validated data is assigned to a variable before reaching a sink.
+ * Per project principle "silence over false positives", this is acceptable.
+ *
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:31]: Added upstream link (vimeo/psalm#11765) to ValidatedTypeHandler docblock.

File: stubs/common/Foundation/Http/FormRequest.stubphp [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+namespace Illuminate\Foundation\Http;
+
+class FormRequest extends \Illuminate\Http\Request
+{
+    /**
+     * Get the validated data from the request.
+     *
+     * @param  array|int|string|null  $key
+     * @param  mixed  $default
+     * @return ($key is null ? array<string, mixed> : mixed)
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply stubs/common/Foundation/Http/FormRequest.stubphp]: Fixed — conditional return type now handles array keys: `($key is null ? array<string, mixed> : ($key is array ? array<string, mixed> : mixed))`
